src/ChangeStream.php

```diff
@@ -82,15 +85,22 @@ class ChangeStream implements Iterator
      */
     private bool $hasAdvanced = false;
 
+    private ?DocumentCodec $codec;
+
     /**
      * @internal
      *
      * @param ResumeCallable $resumeCallable
      */
-    public function __construct(ChangeStreamIterator $iterator, callable $resumeCallable)
+    public function __construct(ChangeStreamIterator $iterator, callable $resumeCallable, ?DocumentCodec $codec = null)
```

Comment:
Watch constructs the ChangeStreamIterator and has access to its internal cursor, even though it currently sends it as-is directly from the private `executeAggregate()` method. Watch also constructs the `$resumeCallable` here.

I was going to ask why ChangeStream bothers with calling `setTypeMap(['root' => 'bson'])` when that could otherwise be done in Watch; however, even if we _did_ move that logic to Watch, ChangeStream would still need to know about the codec in order to process return values for `current()`.

And you previously explained that Watch cannot apply the codec itself since doing so might interfere with resume token collection.

I think this is fine as-is but just want to confirm with you before signing off. Feel free to resolve after reading if my understanding is correct.

Another reason for asking was that ChangeStream is in the public API (although undocumented), so I'm paying extra attention the signature change above.

Reply:
Since `ChangeStream` needs to be in control of the codec used in the cursor (as it needs to be sure about the return types), I figured it would make more sense for `ChangeStream` to apply its codec to the underlying cursor instead of relying on that being set correctly. As is, `Watch` is only accepting the `codec` option on behalf of `ChangeStream`, and passing it on without needing to know about internal workings of `ChangeStream`.

The main problem with applying the codec in the original `Aggregate` operation is that while the aggregation pipeline results will always have an `_id` field to be used as resume token, we cannot make such guarantees after the codec has been applied. After that, the `_id` field may have been discarded or assigned to an object property we don't know about.

Also note that while `ChangeStream` and `ChangeStreamIterator` are part of the public API, both their constructors are marked as internal; indicating that while we expect users to interact with these classes, we don't expect them to instantiate them on their own. The signature change here is backward compatible, as the `$codec` parameter defaults to `null`, so any previously working instantiations of that class will continue to function as expected.